### PR TITLE
Better revision selection and testing for incremental check/ tools.

### PR DESCRIPTION
**Pull request by Strilanc**
- If no revision is specified, try a few defaults.
- Fail fast if the revision doesn't exist.

### CODE REVIEW

File: check/pytest-and-incremental-coverage [hunk 1]
```diff
@@ -19,7 +28,28 @@ cd ${own_directory}
...
 
-PYTHONPATH="$(pwd)" python dev_tools/run_pytest_and_incremental_coverage.py $@
+# Figure out which branch to compare against.
+rest="$@"
+if [ ! -z "$1" ] && [[ $1 != -* ]]; then
+    branch="$1"
+    rest="${@:2}"
+elif [ ! -z $(git branch --list upstream/master) ]; then
+    branch=upstream/master
+elif [ !-z $(git branch --list origin/master) ]; then
```
Comment (cduck) [inline check/pytest-and-incremental-coverage]: Needs a space between ! and -z.

File: check/pytest-and-incremental-coverage [hunk 2]
```diff
@@ -19,7 +28,28 @@ cd ${own_directory}
...
+    branch=origin/master
+elif [ !-z $(git branch --list master) ]; then
```
Comment (cduck) [inline check/pytest-and-incremental-coverage]: Also missing a space.

File: check/pytest-and-incremental-coverage [hunk 3]
```diff
@@ -19,7 +28,28 @@ cd ${own_directory}
 repo_dir=$(git rev-parse --show-toplevel)
 cd ${repo_dir}
 
-PYTHONPATH="$(pwd)" python dev_tools/run_pytest_and_incremental_coverage.py $@
+# Figure out which branch to compare against.
+rest="$@"
+if [ ! -z "$1" ] && [[ $1 != -* ]]; then
+    branch="$1"
+    rest="${@:2}"
+elif [ ! -z $(git branch --list upstream/master) ]; then
```
Comment (cduck) [inline check/pytest-and-incremental-coverage]: `git branch --list` only seems to work on local branches.  rev-parse could work.

File: check/pytest-changed-files [hunk 4]
```diff
@@ -39,9 +63,9 @@ changed=$(git diff --name-only "$1" \
...
 num_changed=$(echo -e "${changed}" | wc -w)
 
+# Run it.
 echo "Found ${num_changed} differing files with associated tests."
 if [ -z "${changed}" ]; then
     exit 0
 fi
-
-pytest "${@:2}" ${changed}
+pytest "${rest}" ${changed}
```
Comment (cduck) [inline check/pytest-changed-files]: The quotes need to be removed so $rest can be expanded.

File: check/pytest-and-incremental-coverage [hunk 5]
```diff
@@ -19,7 +28,28 @@ cd ${own_directory}
...
+    echo -e "\e[31mNo master branch found to compare against. Argument #1 must be what to diff against (e.g. 'origin/master' or 'HEAD~1').\e[0m"
+    exit 1
+fi
+set -e
+git rev-parse "${branch}" > /dev/null
+set +e
+echo "Comparing against revision '${branch}'."
+
+# Run it.
+PYTHONPATH="$(pwd)" python dev_tools/run_pytest_and_incremental_coverage.py "${branch}" "${rest}"
```
Comment (cduck) [inline check/pytest-and-incremental-coverage]: The quotes need to be removed so $rest can be expanded.

Also, run_pytest_and_incremental_coverage.py seems to ignore all arguments after the first.

Comment (Strilanc) [reply check/pytest-and-incremental-coverage]: Obsoleted.

File: check/pytest-changed-files (hunk 4 above)
Comment (Strilanc) [reply check/pytest-changed-files]: Fixed.

File: check/pytest-and-incremental-coverage (hunk 3 above)
Comment (Strilanc) [reply check/pytest-and-incremental-coverage]: Obsoleted.

File: check/pytest-and-incremental-coverage (hunk 2 above)
Comment (Strilanc) [reply check/pytest-and-incremental-coverage]: Obsoleted.

File: check/pytest-and-incremental-coverage (hunk 1 above)
Comment (Strilanc) [reply check/pytest-and-incremental-coverage]: Obsoleted.

File: check/pytest-and-incremental-coverage [hunk 6]
```diff
@@ -19,7 +28,28 @@ cd ${own_directory}
...
+elif [ !-z $(git branch --list origin/master) ]; then
+    branch=origin/master
+elif [ !-z $(git branch --list master) ]; then
+    branch=master
+else
+    echo -e "\e[31mNo master branch found to compare against. Argument #1 must be what to diff against (e.g. 'origin/master' or 'HEAD~1').\e[0m"
+    exit 1
+fi
+set -e
+git rev-parse "${branch}" > /dev/null
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply check/pytest-and-incremental-coverage]: Done.

**Comment by Strilanc** [comment]
O.k.a.y. I think it's working now. That was quite the journey.

File: dev_tools/bash_scripts_test.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    assert "No revision 'HEAD~9999'." in result.err
+
+    result = run(script_file='check/pytest-changed-files')
+    assert result.exit_code == 0
+    assert result.out == ''
+    assert result.err == ("Comparing against revision 'master'.\n"
+                          "Found 0 differing files with associated tests.\n")
+
+    result = run(script_file='check/pytest-changed-files',
+                 setup='git branch origin/master')
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply dev_tools/bash_scripts_test.py]: I added a test for that.

**Pull request merged**